### PR TITLE
Don't use Fast Fetch with AdSense or DoubleClick on non-Google AMP caches

### LINKED ISSUE #10871: AdSense and DoubleClick try to Fast Fetch on non-Google AMP caches
This is a problem because the AdSense and DoubleClick ad servers are currently configured to reject CORS requests that aren't from subdomains of cdn.ampproject.org or localhost. So ads don't show on those caches.

### PR DESCRIPTION
**Pull request by taymonbeal**
Fixes #10871.

### CODE REVIEW

File: extensions/amp-ad-network-adsense-impl/0.1/test/test-adsense-a4a-config.js [hunk 1]
```diff
@@ -63,6 +64,19 @@ describe('adsense-a4a-config', () => {
       expect(adsenseIsA4AEnabled(mockWin, elem)).to.be.false;
     });
 
+    it('should not enable a4a when on a non-Google AMP cache', () => {
+      mockWin.location = parseUrl(
+          'https://amp.cloudflare.com/some/path/to/content.html');
+      sandbox.stub(
+          urls, 'cdnProxyRegex',
```
Comment (erwinmombay) [inline extensions/amp-ad-network-adsense-impl/0.1/test/test-adsense-a4a-config.js]: this works in practice but shouldn't be done as modules are not really objects and can't be stubbed

Comment (taymonbeal) [inline extensions/amp-ad-network-adsense-impl/0.1/test/test-adsense-a4a-config.js]: urls isn't a module, it's just an object.

Comment (erwinmombay) [inline extensions/amp-ad-network-adsense-impl/0.1/test/test-adsense-a4a-config.js]: ah i stand corrected

File: ads/google/a4a/utils.js [hunk 2]
```diff
@@ -88,13 +88,10 @@ export const TRUNCATION_PARAM = {name: 'trunc', value: '1'};
...
-  // Note: Theoretically, isProxyOrigin is the right way to do this, b/c it
-  // will be kept up to date with known proxies.  However, it doesn't seem to
-  // be compatible with loading the example files from localhost.  To hack
-  // around that, just say that we're A4A eligible if we're in local dev
-  // mode, regardless of origin path.
+  const googleCdnProxyRegex =
+      /^https:\/\/([a-zA-Z0-9_-]+\.)?cdn\.ampproject\.org/;
   return supportsNativeCrypto &&
-      (isProxyOrigin(win.location) || getMode(win).localDev ||
+      (googleCdnProxyRegex.test(win.location.origin) || getMode(win).localDev ||
```
Comment (eshienbrood) [inline ads/google/a4a/utils.js]: I'm not seeing how this is functionally different from what `isProxyOrigin` does, i.e., it's basically
`return urls.cdnProxyRegex.test(url.origin);` where` urls.cdnProxyRegex` is the same as what you've added above (unless overridden).

Comment (taymonbeal) [inline ads/google/a4a/utils.js]: "Unless overridden" is exactly the scenario where the problem manifests. `urls.cdnProxyRegex` can be overridden in `AMP_CONFIG`, and Cloudflare does so.

Comment (eshienbrood) [inline ads/google/a4a/utils.js]: Ah, OK.

Comment (taymonbeal) [inline ads/google/a4a/utils.js]: To see for yourself, search for the following string in https://amp.cloudflare.com/v0.js: `"cdnProxyRegex":/^https:\/\/([a-zA-Z0-9_-]+\.)?amp\.cloudflare\.com/,`

Comment (jonkeller) [inline ads/google/a4a/utils.js]: Nit: this will match https://cdn.ampproject.org.badguys.com/etc

Comment (taymonbeal) [inline ads/google/a4a/utils.js]: Good catch. We can't add a trailing slash though, that won't match an origin; we need a trailing $. Also, that regex was copy-pasted from `isProxyOrigin` so other code has the same bug.

**Comment by aghassemi** [comment]
@erwinmombay @lannka Was this also patched into Canary (1502234866079)?

**Pull request merged**